### PR TITLE
Fix C++ unit tests

### PR DESCRIPTION
**Pull request by lebeg**
## Description ##
Fixes the C++ unit tests. The operation registrations were not invoked due to symbol stripping by the linker on Mac (clang).

## Dependencies ##

This change depends on:

- apache:master -> **Fix broken unit test compilation** [#9321](https://github.com/apache/incubator-mxnet/pull/9321)
- dmlc:master -> **Add fully qualified type on define default_real_t** [#317](https://github.com/dmlc/mshadow/pull/317/commits)

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [Removed tabs from cmake file](https://github.com/apache/incubator-mxnet/pull/9349/commits/23d04ff84f239ca5e43c348c571d6e70a37dc446)
- [Added special clang flag](https://github.com/apache/incubator-mxnet/pull/9349/commits/e1fd101ae0f304b6e5a03ad604ac69cd3d2fbf4b)

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -20,6 +20,11 @@ if(NOT MSVC)
   set(UNITTEST_STATIC_LINK ON)
 endif()
 
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
```
Comment (cjolivier01) [inline tests/CMakeLists.txt]: This should be part of the BEGIN_WHOLE_ARCHIVE/E#ND_WHOLE_ARCHIVE definition in the top-level CMakeLists.txt

Comment (lebeg) [reply tests/CMakeLists.txt]: Sure, done.

**Comment by lebeg** [comment]
The verification requires on dmlc:master 'Add fully qualified type on define default_real_t' #317 to be merged first.

**Comment by lebeg** [comment]
@cjolivier01 Could you take another look please?

File: CMakeLists.txt [hunk 2]
```diff
@@ -201,9 +201,13 @@ include_directories("dlpack/include")
 #  add_subdirectory(dlpack)
 #endif()
 
+# Prevent stripping out symbols (operator registrations, for example)
 if(NOT MSVC AND NOT APPLE)
   set(BEGIN_WHOLE_ARCHIVE -Wl,--whole-archive)
   set(END_WHOLE_ARCHIVE -Wl,--no-whole-archive)
+elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  # using regular Clang or AppleClang
+  set(BEGIN_WHOLE_ARCHIVE -Wl,-force_load)
```
Comment (cjolivier01) [inline CMakeLists.txt]: nit: in the interest of completeness and readability, it would be nice to have a set(END_WHOLE_ARCHIVE "") here

**Pull request merged**